### PR TITLE
Update staking v2

### PR DESCRIPTION
**Pull request by duypv1997**
- Implement params set contract
- Implement proposal and vote to change max validator
- Remove min self delegation
- Add signing info and modify validator APIs
- Repo staking v2: https://github.com/kardiachain/staking-contract/pull/8

### CODE REVIEW

File: kvm/smc/dpos/Params.sol [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            minSignedPerWindow: 5 * 10**16,
+            minStake: 25000 * 10**18, // 25 000 KAI
+            minValidatorBalance: 12500000 * 10**18 // 12500000 KAI
+        });
+
+        mintParams = MintParams({
+            inflationRateChange: 2 * 10**16, // 2%
+            goalBonded: 35 * 10**16,
+            blocksPerYear: 6307200,
+            inflationMax: 7 * 10**16,
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:67]: 5%

File: kvm/smc/dpos/Params.sol [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            inflationMin: 189216 * 10**11 // 1,89216%
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:68]: 1%

File: kvm/smc/dpos/Params.sol [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            unbondingTime: 1814400, 
+            slashFractionDoubleSign: 5 * 10**16,
+            signedBlockWindow: 10000,
+            minSignedPerWindow: 5 * 10**16,
+            minStake: 25000 * 10**18, // 25 000 KAI
+            minValidatorBalance: 12500000 * 10**18 // 12500000 KAI
+        });
+
+        mintParams = MintParams({
+            inflationRateChange: 2 * 10**16, // 2%
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:64]: 1%

File: kvm/smc/dpos/Params.sol [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            goalBonded: 35 * 10**16,
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:65]: 50%

File: kvm/smc/dpos/Params.sol [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    MintParams public mintParams;
+
+    constructor() public {
+        transferOwnership(msg.sender);
+        _staking = IStaking(msg.sender);
+
+        stakingParams = StakingParams({
+            baseProposerReward: 1 * 10**16,
+            bonusProposerReward: 4 * 10**16,
+            maxValidators: 20
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:49]: maxProposers

File: kvm/smc/dpos/Params.sol [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+        validatorParams = ValidatorParams({
+            downtimeJailDuration: 259200,
+            slashFractionDowntime: 1 * 10**14,
+            unbondingTime: 1814400, 
+            slashFractionDoubleSign: 5 * 10**16,
+            signedBlockWindow: 10000,
+            minSignedPerWindow: 5 * 10**16,
+            minStake: 25000 * 10**18, // 25 000 KAI
+            minValidatorBalance: 12500000 * 10**18 // 12500000 KAI
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:60]: minValidatorStake

File: kvm/smc/dpos/Params.sol [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            maxValidators: 20
+        });
+
+        validatorParams = ValidatorParams({
+            downtimeJailDuration: 259200,
+            slashFractionDowntime: 1 * 10**14,
+            unbondingTime: 1814400, 
+            slashFractionDoubleSign: 5 * 10**16,
+            signedBlockWindow: 10000,
+            minSignedPerWindow: 5 * 10**16,
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:58]: 50%

File: kvm/smc/dpos/Params.sol [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        stakingParams = StakingParams({
+            baseProposerReward: 1 * 10**16,
+            bonusProposerReward: 4 * 10**16,
+            maxValidators: 20
+        });
+
+        validatorParams = ValidatorParams({
+            downtimeJailDuration: 259200,
+            slashFractionDowntime: 1 * 10**14,
+            unbondingTime: 1814400,
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:55]: 604800 // 7 days for withdraw

File: kvm/smc/dpos/Params.sol [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+            baseProposerReward: 1 * 10**16,
+            bonusProposerReward: 4 * 10**16,
+            maxValidators: 20
+        });
+
+        validatorParams = ValidatorParams({
+            downtimeJailDuration: 259200,
+            slashFractionDowntime: 1 * 10**14,
+            unbondingTime: 1814400, 
+            slashFractionDoubleSign: 5 * 10**16,
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:56]: 25%

File: kvm/smc/dpos/Params.sol [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+        stakingParams = StakingParams({
+            baseProposerReward: 1 * 10**16,
+            bonusProposerReward: 4 * 10**16,
+            maxValidators: 20
+        });
+
+        validatorParams = ValidatorParams({
+            downtimeJailDuration: 259200,
+            slashFractionDowntime: 1 * 10**14,
```
Comment (lewtran) [inline kvm/smc/dpos/Params.sol:54]: 5%

**Review by lewtran — CHANGES_REQUESTED** [review]

**Comment by lewtran** [comment]
Please update the latest staking source code

File: mainchain/api_interface.go [hunk 11]
```diff
@@ -180,7 +180,7 @@ func (k *KardiaService) GetValidators() ([]*staking.Validator, error) {
 		}
 		valInfo, err := k.validator.GetInforValidator(st, header, k.blockchain, kvmConfig, valContractAddr)
 		if err != nil {
-			return nil, err
+			continue
```
Comment (lewtran) [inline mainchain/api_interface.go:183]: double check this @duypv1997

**Review by lewtran — CHANGES_REQUESTED** [review]

File: mainchain/genesis/genesis.go [hunk 12]
```diff
@@ -349,16 +349,10 @@ func setupGenesisStaking(staking *staking.StakingSmcUtil, statedb *state.StateDB
...
-			return fmt.Errorf("apply create validator err: %s", err)
-		}
-
-		if err := staking.StartGenesisValidator(statedb, header, nil, cfg,
-			common.HexToAddress(val.Address),
+			val.MaxChangeRate,
 			val.SelfDelegate); err != nil {
-			return fmt.Errorf("apply start validator err: %s", err)
+			return fmt.Errorf("apply create validator err: %s", err)
 		}
```
Comment (lewtran) [inline mainchain/genesis/genesis.go:355]: Re enable this

**Review by lewtran — CHANGES_REQUESTED** [review]

**Comment by lewtran** [comment]
https://github.com/kardiachain/go-kardia/blob/Update/staking_v2/kai/state/cstate/store.go#L323 
Implement here to check if the validators start with genesis or not

**Comment by lewtran** [comment]
Add "StartWithGenesis" flag
https://github.com/kardiachain/go-kardia/blob/Update/staking_v2/cmd/cfg/genesis.yaml#L31

File: mainchain/genesis/genesis.go (hunk 12 above)
Comment (y4t0g4m1) [reply mainchain/genesis/genesis.go:356]: Done!

**Pull request merged**